File: test/ablation-03/03-ablation-biology.R
#!/usr/bin/env Rscript
# Biological-anchor readout for ablation-03. External RDS inputs are read-only.

options(stringsAsFactors = FALSE)
.ablation03_find_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  candidates <- c(
    if (length(file_arg)) dirname(sub("^--file=", "", file_arg[1L])) else character(),
    getwd()
  )
  for (candidate in unique(candidates)) {
    candidate <- normalizePath(candidate, winslash = "/", mustWork = FALSE)
    if (file.exists(file.path(candidate, "ablation-03.Rproj"))) return(candidate)
    nested <- file.path(candidate, "test", "ablation-03")
    if (file.exists(file.path(nested, "ablation-03.Rproj"))) return(normalizePath(nested, winslash = "/"))
  }
  stop("ablation-03: cannot locate ablation-03.Rproj.", call. = FALSE)
}
.ablation03_dir <- .ablation03_find_dir()
root <- normalizePath(file.path(.ablation03_dir, "..", ".."), winslash = "/", mustWork = TRUE)
out_dir <- file.path(.ablation03_dir, "tmp", "ablation-biology")
fig_dir <- file.path(.ablation03_dir, "figures")
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
dir.create(fig_dir, recursive = TRUE, showWarnings = FALSE)

result_dir <- file.path(.ablation03_dir, "tmp", "ablation-experiment")
manifest <- readRDS(file.path(result_dir, "manifest.rds"))
retrieval <- readRDS(file.path(result_dir, "retrieval.rds"))
neighbours <- retrieval$neighbors[retrieval$neighbors$neighbor_rank <= 15, , drop = FALSE]
target_ids <- sort(unique(c(as.character(neighbours$query_sample), as.character(neighbours$reference_sample))))
full_path <- Sys.getenv("CCS_FULL_EXPRESSION_RDS", unset = "E:/Sync/@Analysis/PanCan_Data/Level 1/PanCan_CancerSample_DataListForCCS_GEO+cBioPortal+UCXCXenav20240809.rds")
sig_path <- Sys.getenv("CCS_GENE_SIGNATURE_RDS", unset = "E:/RCloud/database/Signature/report/GeneSignature-HWB.rds")
cache_path <- Sys.getenv("CCS_BIOLOGY_CACHE_RDS", unset = file.path(out_dir, "expression-anchor-cache.rds"))
if (!file.exists(cache_path)) {
  stop("ablation-03 biology: expression-anchor-cache.rds is missing; run 01-ablation03-biology-cache.R first.", call. = FALSE)
}
cache <- readRDS(cache_path)
if (!identical(cache$schema_version, 1L) || !identical(cache$status, "complete")) {
  stop("ablation-03 biology: unsupported or incomplete cache schema.", call. = FALSE)
}
expected_sample_hash <- digest::digest(paste(target_ids, collapse = "\n"), algo = "md5", serialize = FALSE)
if (!identical(cache$sample_key_hash, expected_sample_hash)) {
  stop("ablation-03 biology: cache sample-key hash mismatch; rebuild the cache.", call. = FALSE)
}
if (!identical(normalizePath(sig_path, winslash = "/", mustWork = FALSE), cache$signature$path)) {
  stop("ablation-03 biology: signature source mismatch; rebuild the cache.", call. = FALSE)
}
if (file.exists(full_path)) {
  source_hash <- digest::digest(file = full_path, algo = "md5")
  if (!identical(source_hash, cache$source$md5)) {
    stop("ablation-03 biology: expression source hash mismatch; rebuild the cache.", call. = FALSE)
  }
}
source(file.path(.ablation03_dir, "03-ablation-biology_functions.R"))
anchors <- cache$anchors
coverage <- cache$coverage
coverage$external_query_cohort <- coverage$cohort_key %in% manifest$external_cohorts
# Fit one gene-wise reference scale across all non-query cohorts, then apply
# that fixed transform to both query and reference samples.  Per-cohort z-scoring
# would put every cohort in a different coordinate system and make cross-cohort
# absolute deltas uninterpretable.
reference_keys <- setdiff(names(cache$cohorts), manifest$external_cohorts)
if (!length(reference_keys)) {
  stop("ablation-03 biology: no reference cohorts available for global scaling.", call. = FALSE)
}
global_stats <- lapply(sort(unique(unlist(anchors, use.names = FALSE))), function(gene) {
  values <- unlist(lapply(cache$cohorts[reference_keys], function(cohort) {
    mat <- cohort$expression
    if (!gene %in% rownames(mat)) return(numeric())
    as.numeric(mat[gene, , drop = TRUE])
  }), use.names = FALSE)
  values <- values[is.finite(values)]
  if (length(values) < 2L) return(c(mean = NA_real_, sd = NA_real_, n = length(values)))
  c(mean = mean(values), sd = stats::sd(values), n = length(values))
})
global_stats <- do.call(rbind, global_stats)
rownames(global_stats) <- sort(unique(unlist(anchors, use.names = FALSE)))
global_stats <- as.data.frame(global_stats, stringsAsFactors = FALSE)
global_stats$gene_id <- rownames(global_stats)
global_stats <- global_stats[is.finite(global_stats$mean) & is.finite(global_stats$sd) &
  global_stats$sd > 0, , drop = FALSE]
score_rows <- list(); ii <- 0L
for (cohort in cache$cohorts) {
  mat <- cohort$expression
  ids <- cohort$sample_id
  for (anchor in names(anchors)) {
    genes <- intersect(anchors[[anchor]], intersect(rownames(mat), global_stats$gene_id))
    if (length(genes) < 2L) next
    stats <- global_stats[match(genes, global_stats$gene_id), , drop = FALSE]
    values <- mat[genes, , drop = FALSE]
    z <- sweep(values, 1L, stats$mean, FUN = "-")
    z <- sweep(z, 1L, stats$sd, FUN = "/")
    score <- colMeans(z, na.rm = TRUE)
    ii <- ii + 1L
    score_rows[[ii]] <- data.frame(sample_id = ids, anchor = anchor,
      score = as.numeric(score), stringsAsFactors = FALSE)
  }
}
if (!length(score_rows)) stop("ablation-03 biology: no anchor has at least two cached genes.", call. = FALSE)
scores <- do.call(rbind, score_rows)
scores <- scores[is.finite(scores$score), , drop = FALSE]

boot_ci <- function(x, seed = 20260830L, B = 500L) {
  x <- x[is.finite(x)]; if (!length(x)) return(c(mean = NA_real_, low = NA_real_, high = NA_real_))
  set.seed(seed); draws <- replicate(B, mean(sample(x, length(x), replace = TRUE)))
  c(mean = mean(x), low = unname(quantile(draws, .025)), high = unname(quantile(draws, .975)))
}
utility_rows <- list(); kk <- 0L
per_query_rows <- list(); pp <- 0L
missing_rows <- list(); mm <- 0L
for (anchor in names(anchors)) {
  sc <- scores[scores$anchor == anchor, c("sample_id", "score")]
  q <- merge(neighbours, sc, by.x = "query_sample", by.y = "sample_id")
  names(q)[names(q) == "score"] <- "query_score"
  q <- merge(q, sc, by.x = "reference_sample", by.y = "sample_id")
  names(q)[names(q) == "score"] <- "reference_score"
  q$abs_delta <- abs(q$query_score - q$reference_score)
  q$utility <- exp(-q$abs_delta)
  per_query <- aggregate(cbind(abs_delta, utility) ~ representation + query_sample + query_cohort,
    data = q, FUN = mean)
  per_query$anchor <- anchor
  pp <- pp + 1L
  per_query_rows[[pp]] <- per_query[, c(
    "anchor", "representation", "query_sample", "query_cohort", "utility"
  )]
  for (rep in unique(per_query$representation)) {
    d <- per_query[per_query$representation == rep, , drop = FALSE]
    ci_u <- boot_ci(d$utility, seed = 20260830L + kk)
    ci_d <- boot_ci(d$abs_delta, seed = 20300830L + kk)
    kk <- kk + 1L
    expected_pairs <- sum(neighbours$representation == rep)
    valid_pairs <- sum(q$representation == rep)
    utility_rows[[kk]] <- data.frame(anchor = anchor, representation = rep,
      query_count = nrow(d), neighbour_pairs = valid_pairs,
      expected_neighbour_pairs = expected_pairs,
      missing_score_pairs = expected_pairs - valid_pairs,
      mean_abs_delta = ci_d[["mean"]], abs_delta_ci_low = ci_d[["low"]],
      abs_delta_ci_high = ci_d[["high"]], utility = ci_u[["mean"]],
      utility_ci_low = ci_u[["low"]], utility_ci_high = ci_u[["high"]])
    mm <- mm + 1L
    missing_rows[[mm]] <- data.frame(
      anchor = anchor, representation = rep,
      expected_neighbour_pairs = expected_pairs,
      valid_neighbour_pairs = valid_pairs,
      missing_score_pairs = expected_pairs - valid_pairs,
      stringsAsFactors = FALSE
    )
  }
}
utility <- do.call(rbind, utility_rows)
per_query_utility <- do.call(rbind, per_query_rows)
missing_pairs <- do.call(rbind, missing_rows)
anchor_inference <- .biology_paired_contrast(
  per_query_utility,
  n_boot = 2000L,
  seed = 20260830L
)
direct <- utility[utility$representation == "Direct-GSClassifier", , drop = FALSE]
d1 <- utility[utility$representation == "Cohort-d1", , drop = FALSE]
contrasts <- merge(d1, direct, by = "anchor", suffixes = c("_d1", "_direct"))
contrasts <- contrasts[, c("anchor", "mean_abs_delta_d1", "mean_abs_delta_direct", "utility_d1", "utility_direct")]
contrasts$d1_minus_direct_abs_delta <- contrasts$mean_abs_delta_d1 - contrasts$mean_abs_delta_direct
contrasts$d1_minus_direct_utility <- contrasts$utility_d1 - contrasts$utility_direct
contrasts$interpretation <- ifelse(contrasts$d1_minus_direct_utility > 0, "d1 higher utility", "Direct higher utility")

write.csv(coverage, file.path(out_dir, "anchor_coverage.csv"), row.names = FALSE)
write.csv(utility, file.path(out_dir, "anchor_utility.csv"), row.names = FALSE)
write.csv(missing_pairs, file.path(out_dir, "anchor_missing_pairs.csv"), row.names = FALSE)
write.csv(contrasts, file.path(out_dir, "anchor_contrasts.csv"), row.names = FALSE)
write.csv(anchor_inference, file.path(out_dir, "anchor_inference.csv"), row.names = FALSE)
saveRDS(list(anchors = anchors, coverage = coverage, utility = utility, contrasts = contrasts,
             inference = anchor_inference,
             retrieval_rows_top15 = nrow(neighbours), source_signature = sig_path,
             cache_path = cache_path, cache_schema_version = cache$schema_version,
             cache_source_md5 = cache$source$md5, cache_sample_key_hash = cache$sample_key_hash,
             scaling = list(method = "reference_cohort_global_gene_zscore",
                            reference_cohort_count = length(reference_keys),
                            reference_cohorts = reference_keys,
                            genes_with_finite_scale = nrow(global_stats))),
        file.path(out_dir, "ablation03-biology.rds"))

# Figures are rendered centrally by 02-ablation03-experiment.Rmd so that the
# HTML, PDF, and preview outputs always share one current data source.  This
# stage writes tabular/RDS products only and does not recreate deprecated
# biological-anchor figures under names that could be mistaken for report plots.
cat(sprintf("anchors=%d coverage_rows=%d utility_rows=%d output=%s\n", length(anchors), nrow(coverage), nrow(utility), out_dir))

File: test/ablation-03/03-ablation-biology.py
#!/usr/bin/env python3
"""Read-only biological-anchor analysis for ablation-03.

The expression atlas and signature RDS files are never written.  Outputs are
created only below ``test/ablation-03``.  This Python fallback is used when the
host does not provide R/Rscript; the calculations mirror the planned
cross-cohort anchor readout (per-cohort z-scored signature means and paired
Direct/d1 neighbour comparisons).
"""
from __future__ import annotations

import json
import hashlib
import math
import os
import re
from pathlib import Path

import numpy as np
import pandas as pd
import rdata


ROOT = Path(__file__).resolve().parents[2]
OUT = ROOT / "test" / "ablation-03" / "tmp" / "ablation-biology"
FIG = ROOT / "test" / "ablation-03" / "figures"
OUT.mkdir(parents=True, exist_ok=True)
FIG.mkdir(parents=True, exist_ok=True)

SIG_RDS = Path(os.getenv(
    "CCS_GENE_SIGNATURE_RDS",
    r"E:\RCloud\database\Signature\report\GeneSignature-HWB.rds",
))
FULL_RDS = Path(os.getenv("CCS_FULL_EXPRESSION_RDS", "")) if os.getenv("CCS_FULL_EXPRESSION_RDS") else None
CACHE_RDS = Path(os.getenv("CCS_BIOLOGY_CACHE_RDS", str(OUT / "expression-anchor-cache.rds")))
RETRIEVAL_RDS = ROOT / "test" / "ablation-03" / "tmp" / "ablation-experiment" / "retrieval.rds"
MANIFEST_RDS = ROOT / "test" / "ablation-03" / "tmp" / "ablation-experiment" / "manifest.rds"


def _scalar(value):
    if isinstance(value, np.ndarray):
        if value.size == 1:
            return _scalar(value.reshape(-1)[0])
        return [_scalar(v) for v in value.reshape(-1)]
    if isinstance(value, (np.generic,)):
        return value.item()
    return value


def _flatten_genes(value):
    genes = []
    if isinstance(value, dict):
        for child in value.values():
            genes.extend(_flatten_genes(child))
    elif isinstance(value, np.ndarray):
        genes.extend(str(v) for v in value.reshape(-1))
    elif isinstance(value, (list, tuple)):
        for child in value:
            genes.extend(_flatten_genes(child))
    return sorted({g for g in genes if g and g != "nan"})


def _find_signature(signatures, predicate):
    for family, entries in signatures.items():
        if not isinstance(entries, dict):
            continue
        for name, value in entries.items():
            name_s = str(name)
            if predicate(name_s):
                genes = _flatten_genes(value)
                if genes:
                    return str(family), name_s, genes
    return None


def _zscore_score(values, gene_idx):
    sub = np.asarray(values[gene_idx, :], dtype=float)
    mu = np.nanmean(sub, axis=1, keepdims=True)
    sd = np.nanstd(sub, axis=1, keepdims=True)
    sd[~np.isfinite(sd) | (sd == 0)] = np.nan
    z = (sub - mu) / sd
    return np.nanmean(z, axis=0)


def _bootstrap_mean(values, rng, n=500):
    values = np.asarray(values, dtype=float)
    values = values[np.isfinite(values)]
    if values.size == 0:
        return (math.nan, math.nan, math.nan)
    draws = values[rng.integers(0, values.size, size=(n, values.size))].mean(axis=1)
    return float(values.mean()), float(np.quantile(draws, 0.025)), float(np.quantile(draws, 0.975))


def _svg_to_deliverables(svg_text, stem):
    """Write SVG and convert it to the required PDF/JPG previews."""
    svg_path = FIG / f"{stem}.svg"
    svg_path.write_text(svg_text, encoding="utf-8")
    import subprocess
    subprocess.run(["magick", str(svg_path), str(FIG / f"{stem}.pdf")], check=True,
                   capture_output=True)
    subprocess.run(["magick", "-density", "220", str(svg_path), str(FIG / f"{stem}.jpg")],
                   check=True, capture_output=True)


def _bar_svg(labels, values, x_label, stem, width=720, height=420, horizontal=False, series=None):
    margin = {"l": 170 if horizontal else 70, "r": 25, "t": 25, "b": 90}
    plot_w, plot_h = width - margin["l"] - margin["r"], height - margin["t"] - margin["b"]
    colors = ["#4878a8", "#c85c5c"]
    out = [f'<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}">',
           '<rect width="100%" height="100%" fill="white"/>', '<g font-family="Arial, sans-serif" font-size="12" fill="#222">']
    if horizontal:
        maxv = max(1.0, max(values) * 1.08)
        for i, (lab, val) in enumerate(zip(labels, values)):
            y = margin["t"] + (i + 0.5) * plot_h / len(labels)
            w = plot_w * val / maxv
            out.append(f'<rect x="{margin["l"]}" y="{y-10:.1f}" width="{w:.1f}" height="20" fill="{colors[0]}"/>')
            out.append(f'<text x="{margin["l"]-8}" y="{y+4:.1f}" text-anchor="end">{lab}</text>')
        out.append(f'<text x="{margin["l"]+plot_w/2:.1f}" y="{height-25}" text-anchor="middle">{x_label}</text>')
    else:
        maxv = max(1.0, max(values) * 1.08)
        for i, (lab, val) in enumerate(zip(labels, values)):
            x = margin["l"] + (i + 0.5) * plot_w / len(labels)
            h = plot_h * val / maxv
            out.append(f'<rect x="{x-18:.1f}" y="{margin["t"]+plot_h-h:.1f}" width="36" height="{h:.1f}" fill="{colors[0]}"/>')
            out.append(f'<text x="{x:.1f}" y="{margin["t"]+plot_h+18}" text-anchor="end" transform="rotate(-25 {x:.1f},{margin["t"]+plot_h+18})">{lab}</text>')
        out.append(f'<text x="{margin["l"]+plot_w/2:.1f}" y="{height-20}" text-anchor="middle">{x_label}</text>')
    out.append('</g></svg>')
    _svg_to_deliverables("".join(out), stem)


def _write_not_estimable(reason, manifest_external=None, retrieval_rows=0):
    """Persist an explicit no-result contract instead of fabricating scores."""
    pd.DataFrame(columns=[
        "tissue", "cohort", "cohort_key", "anchor", "signature",
        "genes_required", "genes_found", "coverage", "sample_count",
        "external_query_cohort", "status", "reason",
    ]).to_csv(OUT / "anchor_coverage.csv", index=False)
    pd.DataFrame(columns=[
        "anchor", "representation", "query_count", "neighbour_pairs",
        "mean_abs_delta", "abs_delta_ci_low", "abs_delta_ci_high",
        "utility", "utility_ci_low", "utility_ci_high", "status", "reason",
    ]).to_csv(OUT / "anchor_utility.csv", index=False)
    pd.DataFrame(columns=[
        "anchor", "d1_minus_direct_abs_delta", "d1_minus_direct_utility",
        "interpretation", "status", "reason",
    ]).to_csv(OUT / "anchor_contrasts.csv", index=False)
    with open(OUT / "anchor_manifest.json", "w", encoding="utf-8") as handle:
        json.dump({
            "status": "not_estimable", "reason": reason,
            "external_cohort_count": int(len(manifest_external or [])),
            "retrieval_rows_top15": int(retrieval_rows),
            "source": str(SIG_RDS),
        }, handle, ensure_ascii=False, indent=2)


def _md5_file(path):
    digest = hashlib.md5()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def main():
    manifest = rdata.read_rds(str(MANIFEST_RDS))
    retrieval = rdata.read_rds(str(RETRIEVAL_RDS))
    neighbors = retrieval["neighbors"].copy()
    neighbors.columns = [str(c) for c in neighbors.columns]
    neighbors["neighbor_rank"] = pd.to_numeric(neighbors["neighbor_rank"])
    neighbors = neighbors[neighbors["neighbor_rank"] <= 15].copy()
    manifest_external = [_scalar(v) for v in np.asarray(manifest["external_cohorts"]).reshape(-1)]
    manifest_external = {str(v) for v in manifest_external}
    target_ids = set(neighbors["query_sample"].astype(str)) | set(neighbors["reference_sample"].astype(str))

    if not CACHE_RDS.exists():
        _write_not_estimable("expression_anchor_cache_missing", manifest_external, len(neighbors))
        raise RuntimeError(f"expression-anchor cache is missing: {CACHE_RDS}")
    try:
        cache = rdata.read_rds(str(CACHE_RDS))
    except Exception as exc:
        _write_not_estimable(f"expression_anchor_cache_unreadable:{type(exc).__name__}", manifest_external,
                             len(neighbors))
        raise RuntimeError("expression-anchor cache is unreadable") from exc
    if int(_scalar(cache.get("schema_version", -1))) != 1 or str(_scalar(cache.get("status", ""))) != "complete":
        raise RuntimeError("expression-anchor cache has unsupported or incomplete schema")
    target_hash = hashlib.md5("\n".join(sorted(target_ids)).encode("utf-8")).hexdigest()
    if str(_scalar(cache.get("sample_key_hash", ""))) != target_hash:
        raise RuntimeError("expression-anchor cache sample-key hash mismatch; rebuild the cache")
    if FULL_RDS is not None and FULL_RDS.exists():
        source = cache.get("source", {})
        expected_md5 = str(_scalar(source.get("md5", ""))) if isinstance(source, dict) else ""
        if expected_md5 and _md5_file(FULL_RDS) != expected_md5:
            raise RuntimeError("expression-anchor cache source hash mismatch; rebuild the cache")
    selected = {}
    raw_anchors = cache.get("anchors", {})
    for label, genes in raw_anchors.items():
        selected[str(label)] = {"genes": _flatten_genes(genes), "name": str(label), "family": "frozen_manifest"}
    if len(selected) < 3:
        raise RuntimeError("Fewer than three frozen signatures are present in expression-anchor cache")

    # Score only samples used by the frozen 02 retrieval table.  Each cohort is
    # standardized separately, preventing platform-wide expression shifts from
    # becoming a biological signal.
    scores = {anchor: {} for anchor in selected}
    coverage_rows = []
    raw_cohorts = cache.get("cohorts", {})
    cohort_items = raw_cohorts.items() if isinstance(raw_cohorts, dict) else enumerate(raw_cohorts)
    for _, arr in cohort_items:
        if not isinstance(arr, dict):
            continue
        tissue = str(_scalar(arr.get("tissue", "")))
        cohort = str(_scalar(arr.get("cohort", "")))
        sample_ids = np.asarray([str(_scalar(v)) for v in np.asarray(arr.get("sample_id", [])).reshape(-1)])
        gene_ids = np.asarray([str(_scalar(v)) for v in np.asarray(arr.get("gene_id", [])).reshape(-1)])
        values = np.asarray(arr.get("expression"), dtype=float)
        if values.ndim != 2 or sample_ids.size == 0:
            continue
        for anchor, spec in selected.items():
            wanted = set(spec["genes"])
            idx = np.flatnonzero(np.isin(gene_ids, list(wanted)))
            coverage_rows.append({
                "tissue": str(tissue), "cohort": cohort, "cohort_key": f"{tissue}/{cohort}",
                "anchor": anchor, "signature": spec["name"], "genes_required": len(wanted),
                "genes_found": int(idx.size), "coverage": float(idx.size / len(wanted)),
                "sample_count": int(sample_ids.size),
                "external_query_cohort": f"{tissue}/{cohort}" in manifest_external,
            })
            if idx.size >= 2:
                score = _zscore_score(values, idx)
                scores[anchor].update({sid: float(v) for sid, v in zip(sample_ids, score) if np.isfinite(v)})
    coverage = pd.DataFrame(coverage_rows)

    # Biological utility is an anchor-profile agreement of retrieved neighbours.
    # ``utility = exp(-|z-score difference|)`` is monotone, bounded, and makes
    # lower distance and higher utility directly comparable across anchors.
    rng = np.random.default_rng(20260830)
    result_rows = []
    for anchor in selected:
        anchor_scores = scores[anchor]
        n = neighbors.copy()
        n["q_score"] = n["query_sample"].astype(str).map(anchor_scores)
        n["r_score"] = n["reference_sample"].astype(str).map(anchor_scores)
        n = n.dropna(subset=["q_score", "r_score"])
        n["abs_delta"] = (n["q_score"] - n["r_score"]).abs()
        n["utility"] = np.exp(-n["abs_delta"])
        grouped = n.groupby(["representation", "query_sample"], as_index=False).agg(
            mean_abs_delta=("abs_delta", "mean"), utility=("utility", "mean"),
            neighbour_pairs=("utility", "size"), query_cohort=("query_cohort", "first"),
        )
        for rep, g in grouped.groupby("representation"):
            mean_u, lo_u, hi_u = _bootstrap_mean(g["utility"], rng)
            mean_d, lo_d, hi_d = _bootstrap_mean(g["mean_abs_delta"], rng)
            result_rows.append({
                "anchor": anchor, "representation": rep, "query_count": int(len(g)),
                "neighbour_pairs": int(g["neighbour_pairs"].sum()),
                "mean_abs_delta": mean_d, "abs_delta_ci_low": lo_d, "abs_delta_ci_high": hi_d,
                "utility": mean_u, "utility_ci_low": lo_u, "utility_ci_high": hi_u,
            })
    utility = pd.DataFrame(result_rows)
    direct = utility[utility.representation == "Direct-GSClassifier"].set_index("anchor")
    d1 = utility[utility.representation == "Cohort-d1"].set_index("anchor")
    contrasts = []
    for anchor in sorted(set(direct.index) & set(d1.index)):
        contrasts.append({
            "anchor": anchor,
            "d1_minus_direct_abs_delta": float(d1.loc[anchor, "mean_abs_delta"] - direct.loc[anchor, "mean_abs_delta"]),
            "d1_minus_direct_utility": float(d1.loc[anchor, "utility"] - direct.loc[anchor, "utility"]),
            "interpretation": "d1 higher utility" if d1.loc[anchor, "utility"] > direct.loc[anchor, "utility"] else "Direct higher utility",
        })
    contrasts = pd.DataFrame(contrasts)

    coverage.to_csv(OUT / "anchor_coverage.csv", index=False)
    utility.reset_index(drop=True).to_csv(OUT / "anchor_utility.csv", index=False)
    contrasts.to_csv(OUT / "anchor_contrasts.csv", index=False)
    with open(OUT / "anchor_manifest.json", "w", encoding="utf-8") as handle:
        json.dump({"anchors": selected, "external_cohort_count": len(manifest_external),
                   "retrieval_rows_top15": int(len(neighbors)), "source": str(SIG_RDS)}, handle, ensure_ascii=False, indent=2)

    # Publication-oriented figures: no in-panel titles; captions live in HTML.
    cov = coverage.groupby("anchor", as_index=False)["coverage"].mean().sort_values("coverage")
    _bar_svg(cov["anchor"].tolist(), cov["coverage"].tolist(),
             "Mean gene coverage across evaluated cohorts", "figure-01-anchor-coverage",
             horizontal=True)
    piv = utility.pivot(index="anchor", columns="representation", values="utility")
    direct_values = piv.get("Direct-GSClassifier", pd.Series(index=piv.index, dtype=float)).fillna(0)
    d1_values = piv.get("Cohort-d1", pd.Series(index=piv.index, dtype=float)).fillna(0)
    # A compact paired SVG is sufficient for the HTML preview; tabular values
    # remain the authoritative result and are written above.
    labels = piv.index.tolist()
    combined = ((direct_values + d1_values) / 2).tolist()
    _bar_svg(labels, combined, "Mean paired anchor utility (Direct/d1)",
             "figure-02-biological-utility", horizontal=False)

    print(json.dumps({"anchors": list(selected), "coverage_rows": len(coverage),
                      "utility_rows": len(utility), "contrast_rows": len(contrasts),
                      "output": str(OUT)}, ensure_ascii=False))


if __name__ == "__main__":
    main()
